src/state.rs

```diff
@@ -299,12 +535,21 @@ impl DemandProxyState {
         };
         let wp_socket_addr = SocketAddr::new(wp_nw_addr.address, gw_address.port);
         match self
-            .fetch_upstream(&wp_nw_addr.network, wp_socket_addr, gw_address.port)
+            .fetch_upstream(&wp_nw_addr.network, wp_socket_addr)
             .await
         {
-            Some(upstream) => {
+            Some(mut upstream) => {
                 debug!(%wl.name, "found waypoint upstream");
-                Ok(Some(upstream))
+                match self
+                    .set_gateway_address(&mut upstream, workload_ip, gw_address.port)
+                    .await
+                {
+                    Ok(_) => Ok(Some(upstream)),
+                    Err(e) => {
```

Comment:
Seems like set_gateway_address has no error paths. Could change the signature and remove branch.

Reply:
`.waypoint_svc_ip_address()?` can return an error in

```rust
pub async fn set_gateway_address(us: &mut Upstream, workload_ip: IpAddr, hbone_port: u16) {
    if us.workload.gateway_address.is_none() {
        us.workload.gateway_address = Some(match us.workload.protocol {
            Protocol::HBONE => {
                let ip = us
                    .workload
                    .waypoint_svc_ip_address()?
                    .unwrap_or(workload_ip);
                SocketAddr::from((ip, hbone_port))
            }
            Protocol::TCP => SocketAddr::from((workload_ip, us.port)),
        });
    }
}
```